src/barnacle/dns.gleam

```diff
@@ -0,0 +1,105 @@
+import gleam/erlang/atom
+import gleam/int
+import gleam/list
+import gleam/result
+import gleam/string
+
+/// A DNS error that can occur when discovering nodes.
+pub type LookupError {
+  FormatError
+  QueryFormatError
+  ServerFailure
+  NoSuchDomain
+  Timeout
+  NotImplemented
+  Refused
+  BadVersion
+  PosixError(String)
+  Unknown
+}
+
+@external(erlang, "barnacle_ffi", "lookup_a")
+fn lookup_a(
+  hostname: String,
+  timeout: Int,
+) -> Result(List(#(Int, Int, Int, Int)), LookupError)
+
+@external(erlang, "barnacle_ffi", "lookup_aaaa")
+fn lookup_aaaa(
+  hostname: String,
+  timeout: Int,
+) -> Result(List(#(Int, Int, Int, Int, Int, Int, Int, Int)), LookupError)
+
+pub type IpAddress {
+  IpV4(Int, Int, Int, Int)
+  IpV6(Int, Int, Int, Int, Int, Int, Int, Int)
+}
+
+/// Perform a DNS lookup against a hostname. This will return a list of IP
+/// addresses. Queries will be performed using the `A` and `AAAA` record types.
+pub fn dns_lookup(
+  hostname: String,
+  timeout: Int,
+) -> Result(List(IpAddress), LookupError) {
+  use a_records <- result.try(lookup_a(hostname, timeout))
+
+  let a_records =
+    a_records
+    |> list.map(fn(ip) {
+      let #(a, b, c, d) = ip
+      IpV4(a, b, c, d)
+    })
+
+  use aaaa_records <- result.try(lookup_aaaa(hostname, timeout))
+
+  let aaaa_records =
+    aaaa_records
+    |> list.map(fn(ip) {
+      let #(a, b, c, d, e, f, g, h) = ip
+      IpV6(a, b, c, d, e, f, g, h)
+    })
+
+  Ok(list.concat([a_records, aaaa_records]))
+}
+
+fn format_ip_address(ip: IpAddress) -> String {
```

Comment:
Do you do this in Gleam instead of calling `inet:ntoa/1` to avoid going back to an Erlang tuple and external ffi?

Reply:
I mostly didn't know that function existed, but I like to keep things in Gleam where possible.